core/src/transport.rs

```diff
@@ -221,149 +221,252 @@ pub trait Transport {
     }
 }
 
-/// Event produced by [`Transport::Listener`]s.
+/// The ID of a single listener.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ListenerId(u64);
+
+impl ListenerId {
+    /// Creates a `ListenerId` from a non-negative integer.
+    pub fn new(id: u64) -> Self {
+        Self(id)
+    }
+}
+
+impl std::ops::Add<u64> for ListenerId {
+    type Output = Self;
+
+    fn add(self, other: u64) -> Self {
+        Self(self.0 + other)
+    }
+}
+
+/// Event produced by [`Transport`]s.
 ///
-/// Transports are expected to produce `Upgrade` events only for
+/// Transports are expected to produce `Incoming` events only for
 /// listen addresses which have previously been announced via
 /// a `NewAddress` event and which have not been invalidated by
 /// an `AddressExpired` event yet.
-#[derive(Clone, Debug, PartialEq)]
-pub enum ListenerEvent<TUpgr, TErr> {
-    /// The transport is listening on a new additional [`Multiaddr`].
-    NewAddress(Multiaddr),
-    /// An upgrade, consisting of the upgrade future, the listener address and the remote address.
-    Upgrade {
-        /// The upgrade.
-        upgrade: TUpgr,
-        /// The local address which produced this upgrade.
+pub enum TransportEvent<T: Transport> {
+    /// A new address is being listened on.
+    NewAddress {
+        /// The listener that is listening on the new address.
+        listener_id: ListenerId,
+        /// The new address that is being listened on.
+        listen_addr: Multiaddr,
+    },
+    /// An address is no longer being listened on.
+    AddressExpired {
+        /// The listener that is no longer listening on the address.
+        listener_id: ListenerId,
+        /// The new address that is being listened on.
+        listen_addr: Multiaddr,
+    },
+    /// A connection is incoming on one of the listeners.
+    Incoming {
+        /// The listener that produced the upgrade.
+        listener_id: ListenerId,
+        /// The produced upgrade.
+        upgrade: T::ListenerUpgrade,
```

Comment:
It would be nice if this were already the final connection.

But that changes a few assumptions. For example, it means that parallelizing connection upgrades would be a concern of the `Transport` implementation and wouldn't be possible as part of `libp2p-swarm`.

On the other hand, it would allow us to simplify the `Transport` trait even more by allowing us to remove another associated type.

Reply:
> But that changes a few assumptions. For example, it means that parallelizing connection upgrades would be a concern of the `Transport` implementation and wouldn't be possible as part of `libp2p-swarm`.

Off the top of my head I prefer decisions on parallelism and concurrency to happen in a single place and not make it a per-transport decision. I think rust-libp2p greatly benefits from the fact that (1) we only spawn one task per connection and (2) everything else is driven by a single task.

Reply:
Yes, I think that is fair. I've actually changed my mind on this. Returning the `Future` here makes more sense and also allows us to move the dialing into this.